modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -387,17 +389,20 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
                     return nodes.get(0);
                 })
                 .thenCompose(sqlNode -> {
+                    final boolean rwTx = dataModificationOp(sqlNode);
+
                     BaseQueryContext ctx = BaseQueryContext.builder()
                             .frameworkConfig(
                                     Frameworks.newConfigBuilder(FRAMEWORK_CONFIG)
                                             .defaultSchema(schema)
-                                            .traitDefs(Commons.LOCAL_TRAITS_SET)
+                                            .traitDefs(rwTx ? Commons.LOCAL_TRAITS_SET : Commons.DISTRIBUTED_TRAITS_SET)
                                             .build()
                             )
                             .logger(LOG)
                             .cancel(queryCancel)
                             .parameters(params)
                             .transaction(outerTx)
+                            .transactionTime(rwTx ? null : new HybridClock().now())
```

Comment:
in case of outer TX we can have different time for few sequented run queries with the same outer TX